crates/sui-core/src/node_sync/node_follower.rs

```diff
@@ -358,66 +379,111 @@ where
+                // we are a validator.
+                let (cert, effects) = self
+                    .download_cert_and_effects(authorities_with_cert, digest, effects_digest)
+                    .await?;
+
+                if is_fullnode {
+                    CertAndEffects::Fullnode(cert, effects)
+                } else {
+                    CertAndEffects::Validator(cert)
+                }
+            }
+        };
 
         // we're done downloading at this point, so we no longer need to prevent other tasks from
         // starting.
         std::mem::drop(permit);
 
-        for parent in effects.effects.dependencies.iter() {
-            let (_, mut rx) = self.pending_txes.wait(parent).await;
+        match cert_and_effects {
+            CertAndEffects::Fullnode(cert, effects) => {
+                for parent in effects.effects.dependencies.iter() {
+                    let (_, mut rx) = self.pending_txes.wait(parent).await;
 
-            if self.state.database.effects_exists(parent)? {
-                continue;
-            }
+                    if self.state.database.effects_exists(parent)? {
+                        continue;
+                    }
 
-            trace!(?parent, digest = ?digests.transaction, "waiting for parent");
-            // Since we no longer hold the semaphore permit, can be sure that our parent will be
-            // able to start.
-            rx.recv()
-                .await
-                .map_err(|e| SuiError::GenericAuthorityError {
-                    error: format!("{:?}", e),
-                })?;
-        }
+                    trace!(?parent, ?digest, "waiting for parent");
+                    // Since we no longer hold the semaphore permit, can be sure that our parent will be
+                    // able to start.
+                    rx.recv()
+                        .await
+                        .map_err(|e| SuiError::GenericAuthorityError {
+                            error: format!("{:?}", e),
+                        })?;
+                }
+
+                if cfg!(debug_assertions) {
+                    for parent in effects.effects.dependencies.iter() {
+                        debug_assert!(self.state.database.effects_exists(parent).unwrap());
+                    }
+                }
+
+                self.state
+                    .handle_node_sync_certificate(cert, effects.clone())
+                    .await?;
 
-        if cfg!(debug_assertions) {
-            for parent in effects.effects.dependencies.iter() {
-                debug_assert!(self.state.database.effects_exists(parent).unwrap());
+                self.effects_stake
+                    .lock()
+                    .unwrap()
+                    .forget_effects(&effects.effects.digest());
+            }
+            CertAndEffects::Validator(cert) => {
+                self.state.handle_certificate(cert).await?;
```

Comment:
This will not work of course if we do not have the dependencies.

Reply:
Correct - an error will be reported to the caller.